# <span style="color:red">Open Hack Session #1</span>

## Today's Agenda
Following Suggestions from: [https://docs.google.com/document/d/1UYKzubjSIA2pcq8qzkzTO62GEqvrCovPzgR7V74cNGk/edit?usp=sharing)
- "I’d like to be able to search for any file from anywhere on my computer"
- "Making code pretty and standardized"
- "How to make pandas latex tables from excel or .csv files"
- "How to use complex numbers in Python?"
- "C++ vs Python and astronomy applications"
- "Brief Intro to Bayesian Statistics"





# Topic 1: ""I’d like to be able to search for any file from anywhere on my computer""

# In this portion there are 2 useful commands to find files within a Command Terminal rather than relying on the user-interface "Finder" application:  ```find```  and  ```grep```

## Set-up for Example:
```
mkdir ~/Desktop/Contact && cd ~/Desktop/Contact
mkdir notes data src .SECRETZ
echo "Hello to Mission Control\nIs anyone there?\n" > notes/transmission_1.txt
echo "MISSION CONTROL?! HELLO?!\n" > notes/shout_1.txt
echo "target=home\n" > src/config.txt
echo "TRANSMISSION_SETTINGS=OFF" >> src/config.txt
echo "Unable to receive confirmation of transmission. ; feeling=confused ; status=ok\n" > data/log1.csv
echo "What if my transmitter was damaged in the storm? I'm unable to request help. ; feeling=scared ; status=not ok\n" > data/log2.csv
echo "I accidentally turned my tranmitting controls off when I sat on the console. Good thing mission control didn't see that..." > .SECRETZ/dont_tell_anyone.oops
```


### Using ```find```: To use find to search for files by name, the general syntax is:
```bash
find <where-to-start> <tests> <actions>
``` 
Let's Practice using it:

- #### Find file by exact name (case-sensitive)
```bash
find . -name "config.txt"
```
produces an output like:
```bash
./src/config.txt
```

- #### Find file by inexact name (case-insensitive) 
```bash
find . -iname "Config.txt"
```
produces an output like:
```bash
./src/config.txt
```

- #### Find by wildcard (e.g., all .txt files in current dirctory):
```bash 
find . -name "*.txt"
```
produces an output like:
```bash
./notes/transmission_1.txt
./notes/shout_1.txt
./src/config.txt
```

- #### Limit how deep in file structure we descend:
```bash
find . -maxdepth 1 -type f -name "*.txt"
``` 
searches only current directory and 
produces an empty output because there aren't any .txt files in `~/Desktop/Contact`


but doing 
```bash
find . -maxdepth 2 -type f -name "*.txt"
``` 
searches the current directory and 1 level of sub-directories within and
produces an output like:
```bash
./notes/transmission_1.txt
./notes/shout_1.txt
./src/config.txt
```


### Using ```grep```: To use grep to search inside files for specific content, the general syntax is:
```bash
grep [OPTIONS] "pattern" file_or_dir
```

#### - Search a single file for 'status' condition:
```bash
grep "status" data/log1.csv
```
produces an output like:
```
Unable to receive confirmation of transmission. ; feeling=confused ; status=ok
```

#### - Recursive search in a folder; show line numbers; ignore case
```bash
grep -Rni "status" .
```

produces an output like:
```
./data/log1.csv:1:Unable to receive confirmation of transmission. ; feeling=confused ; status=ok
./data/log2.csv:1:What if my transmitter was damaged in the storm? I'm unable to request help. ; feeling=scared ; status=not ok
```

#### - Only match whole words (avoid 'roverX' matching)
```bash
grep -Rniw "storm" .
```

produces an output like:
```
./data/log2.csv:1:What if my transmitter was damaged in the storm? I'm unable to request help. ; feeling=scared ; status=not ok
```

#### - Regex (extended) — e.g., match 'status=ok' or 'status=fail'
```bash 
grep -RniE "status=(ok|not ok)" data
```

produces an output like:
```
data/log1.csv:1:Unable to receive confirmation of transmission. ; feeling=confused ; status=ok
data/log2.csv:1:What if my transmitter was damaged in the storm? I'm unable to request help. ; feeling=scared ; status=not ok
```

#### - Knowing what line/phrase to search for might be illuninating

```bash 
grep -Rni "transmission_settings" . 
```

produces an output like:
```
./src/config.txt:3:TRANSMISSION_SETTINGS=OFF
```

And there's the problem...

# Topic #2: "Making code pretty and standardized"
PEP 8, sometimes spelled PEP8 or PEP-8, is the [__official style guide for Python code__](https://peps.python.org/pep-0008/). PEP 8 gives guidelines on naming conventions, code layout, and other best practices. By adhering to PEP 8, you ensure that your Python code is readable and maintainable, which helps with collaboration and professional development.

PEP stands for Python Enhancement Proposal, and there are many [PEPs](https://peps.python.org/). These documents primarily describe new features proposed for the Python language, but some PEPs also focus on design and style and aim to serve as a resource for the community. PEP 8 is one of these style-focused PEPs. Some common PEP 8 example conventions:

| Type | Naming Convention | Examples |
|---|---|---|
| Function | Use a lowercase word or words. <br>Separate words by underscores to improve readability. <br>This style is called snake case. | function, python_function |
| Variable | Use a lowercase single letter, word, or words.<br>Separate words with underscores to improve readability. | x, var, python_variable |
| Class | Start each word with a capital letter. <br>Don’t separate words with underscores. <br>This style is called camel case or Pascal case.	 | Model, PythonClass |
| Method | Use a lowercase word or words. <br>Separate words with underscores to improve readability (snake case). | class_method, method |
| Constant | Use an uppercase single letter, word, or words. <br> Separate words with underscores to improve readability. | CONSTANT, PYTHON_CONSTANT, PYTHON_LONG_CONSTANT |
| Module | Use a short, lowercase word or words. <br>Separate words with underscores to improve readability. | module.py, python_module.py |
| Package | Use a short, lowercase word or words. <br>Don’t separate words with underscores. | package, pythonpackage |


# Python-specific tools for adhering to PEP 8:
- For checking core style/linting:
    - (__Ruff__)[https://docs.astral.sh/ruff/] #<- what we'll use today
    - (__Black__)[https://black.readthedocs.io/en/stable/]
    - (__flake8__)[https://flake8.pycqa.org/en/latest/]
    - (__pycodestyle__)[https://pycodestyle.pycqa.org/en/latest/]
    - (__autopep8__)[https://github.com/hhatto/autopep8]
    - (__Yet Another Python Formatter__)[https://github.com/google/yapf]

# Installing Ruff:
## If using Mini-Forge:
```bash
conda install ruff 
```

## If using Anaconda:
```bash
conda install -c conda-forge ruff 
```

# Example script we'll call `messy.py`:
This code below is very condensed
```python
import sys,os,   math
from collections import  defaultdict
def add(a,b):
    """Return the sum of two numbers.
    
    Parameters
    ----------
    a : float
        First addend.
    b : float
        Second addend.
        
    Returns
    -------
    float
        The sum of ``a`` and ``b``.
    """
    return(a+  b)  
class example:
    """A simple rectangle-like object with ``x`` and ``y`` dimensions.
    
    Parameters
    ----------
    x : float, optional
        First dimension. Default is 1.
    y : float, optional
        Second dimension. Default is 2.
        
    Attributes
    ----------
    x : float
        First dimension.
    y : float
        Second dimension.
    """
  def __init__(self,x= 1,y=2):    
    self.x=x;self.y = y 
  def area (self ):
      """Compute the area as ``x * y``.
      
        Returns
        -------
        float
            Product of ``x`` and ``y``.
        """
    return self.x*self. y
if __name__=="__main__" :
  print('sum:',add(1,2))
  f=example(  3,4)
  print( "area:", f. area())
```

# Applying Ruff to identify PEP 8 no-nos:
```bash
ruff check messy.py --select N
```

This will produce an output like:
```bash
N801 Class name `example` should use CapWords convention
  --> messy.py:17:7
   |
15 |     """
16 |     return(a+  b)  
17 | class example:
   |       ^^^^^^^
18 |     """A simple rectangle-like object with ``x`` and ``y`` dimensions.
19 |     Parameters
   |

Found 1 error.
```


# This will not fix the error for you, just point it out. Typos or non-suggested naming conventions will have to be fixed by you. Let's change our class object name from `example` to `Example` to adhere to PEP 8 rules.

## After this change, let's repeat :
```bash
ruff check messy.py --select N
```

to get this output:
```bash 
All checks passed!
```

# Next, with Ruff we can format our script for a more readable spacing (Linting):
```bash
ruff format messy.py
```

To get this output:
```
1 file reformatted
```

If we then look at our `messy.py` file, we'll see it's been changed to look like:
```python
import sys, os, math
from collections import defaultdict


def add(a, b):
    """Return the sum of two numbers.

    Parameters
    ----------
    a : float
        First addend.
    b : float
        Second addend.

    Returns
    -------
    float
        The sum of ``a`` and ``b``.
    """
    return a + b 


class Example:
    """A simple rectangle-like object with ``x`` and ``y`` dimensions.

    Parameters
    ----------
    x : float, optional
        First dimension. Default is 1.
    y : float, optional
        Second dimension. Default is 2.

    Attributes
    ----------
    x : float
        First dimension.
    y : float
        Second dimension.
    """

    def __init__(self, x=1, y=2):
        self.x = x 
        self.y = y 

    def area(self):
        """Compute the area as ``x * y``.

        Returns
        -------
        float
          Product of ``x`` and ``y``.
        """
        return self.x * self.y
        
if __name__ == "__main__":
    print("sum:", add(1, 2))
    f = Example(3, 4)
    print("area:", f.area())        
```        


# Now our code adheres to PEP 8!

## Note: This isn't perfect. It doesn't format doc-strings or strings, mostly just whitespace (or lack thereof) in code. It also rewrites the input file.

## How to use ruff without overwriting:
```bash
ruff format messy.py --diff
```

#### This will show a line-by-line list of changes that would be reformatted in the input python script. Check the outputs here to verify if this is what you want. Alternatively, you could copy your script before running `ruff format script_name.py` with copy (`cp`), like `cp script_name.py script_name_copy.py `



# Topic #3: "How to make pandas latex tables from excel or .csv files"


In [1]:
import pandas as pd
import numpy as np
# Create a tiny example table
df = pd.DataFrame({
    "Star": ["GJ 1214", "TRAPPIST-1", "Proxima Cen", "Unknown"],
    "Mass (MSun)": [0.176, 0.089, 0.122, None],
    "Radius (RSun)": [0.216, 0.121, 0.154, None],
})
df



,Star,Mass (MSun),Radius (RSun)
0,GJ 1214,0.176,0.216
1,TRAPPIST-1,0.089,0.121
2,Proxima Cen,0.122,0.154
3,Unknown,NaN,NaN


### Quick: get LaTeX code (copy/paste into your .tex file)

In [2]:
latex = df.to_latex(index=False, float_format="%.3f")
print(latex)
print("")

\begin{tabular}{lrr}
\toprule
Star & Mass (MSun) & Radius (RSun) \\
\midrule
GJ 1214 & 0.176 & 0.216 \\
TRAPPIST-1 & 0.089 & 0.121 \\
Proxima Cen & 0.122 & 0.154 \\
Unknown & NaN & NaN \\
\bottomrule
\end{tabular}




### You can modify Alignment per column (l / c/ r for left, center, right):

In [3]:
latex = df.to_latex(index=False, column_format="ccc")
print(latex)
print('')

\begin{tabular}{ccc}
\toprule
Star & Mass (MSun) & Radius (RSun) \\
\midrule
GJ 1214 & 0.176000 & 0.216000 \\
TRAPPIST-1 & 0.089000 & 0.121000 \\
Proxima Cen & 0.122000 & 0.154000 \\
Unknown & NaN & NaN \\
\bottomrule
\end{tabular}




### You can handle missing/NaN values:


In [4]:
latex = df.to_latex(index=False, na_rep="--") #na_rep is NaN replacement character(s)
print(latex)
print('')

\begin{tabular}{lrr}
\toprule
Star & Mass (MSun) & Radius (RSun) \\
\midrule
GJ 1214 & 0.176000 & 0.216000 \\
TRAPPIST-1 & 0.089000 & 0.121000 \\
Proxima Cen & 0.122000 & 0.154000 \\
Unknown & -- & -- \\
\bottomrule
\end{tabular}





### You can modify per column:


In [5]:
latex = df.to_latex(
    index=False,
    formatters={
        "Mass (M☉)": "{:.3f}".format,
        "Radius (R☉)": "{:.3f}".format
    }
)
print(latex)
print('')

\begin{tabular}{lrr}
\toprule
Star & Mass (MSun) & Radius (RSun) \\
\midrule
GJ 1214 & 0.176000 & 0.216000 \\
TRAPPIST-1 & 0.089000 & 0.121000 \\
Proxima Cen & 0.122000 & 0.154000 \\
Unknown & NaN & NaN \\
\bottomrule
\end{tabular}




### you can make long LaTeX tables meant to span a page:


In [6]:
latex_long = df.to_latex(index=False, longtable=True)
print(latex_long)
print("")

\begin{longtable}{lrr}
\toprule
Star & Mass (MSun) & Radius (RSun) \\
\midrule
\endfirsthead
\toprule
Star & Mass (MSun) & Radius (RSun) \\
\midrule
\endhead
\midrule
\multicolumn{3}{r}{Continued on next page} \\
\midrule
\endfoot
\bottomrule
\endlastfoot
GJ 1214 & 0.176000 & 0.216000 \\
TRAPPIST-1 & 0.089000 & 0.121000 \\
Proxima Cen & 0.122000 & 0.154000 \\
Unknown & NaN & NaN \\
\end{longtable}




### you can include multiple commands together


In [7]:
styled_latex = (
    df.style
      .format({"Mass (M☉)": "{:.3f}", "Radius (R☉)": "{:.3f}"})
      .hide(axis='index')          # pandas >= 1.4
      .to_latex(hrules=True, caption="Nearby M-dwarfs (toy example)", label="tab:mdwarfs")
)
print(styled_latex)

\begin{table}
\caption{Nearby M-dwarfs (toy example)}
\label{tab:mdwarfs}
\begin{tabular}{lrr}
\toprule
Star & Mass (MSun) & Radius (RSun) \\
\midrule
GJ 1214 & 0.176000 & 0.216000 \\
TRAPPIST-1 & 0.089000 & 0.121000 \\
Proxima Cen & 0.122000 & 0.154000 \\
Unknown & nan & nan \\
\bottomrule
\end{tabular}
\end{table}




### Alternatively, you can write LaTeX code to file:
```python 
df.to_latex("table_mdwarfs.tex", index=False, float_format="%.3f")
```

### or with style format:
```python
df.style.format({"Mass (M☉)": "{:.3f}", 
                 "Radius (R☉)": "{:.3f}"}).hide(axis='index').to_latex("table_mdwarfs_styled.tex", 
                                                                       hrules=True, 
                                                                       caption="Nearby M-dwarfs (toy example)",
                                                                       label="tab:mdwarfs")
```

# Topic #4: "How to use complex numbers in Python?"
In Python, complex numbers can be defined using the letter `j`:

In [8]:
print( 1, type(1) )
print( 1j, type(1j) )
print( 1+3j, type(1+3j) )
print( (1+3j).real , type( (1+3j).real )  )
print( (1+3j).imag , type( (1+3j).imag )  )

print( abs(1+3j) , type ( abs(1+3j) ) ,'\n')
import numpy as np

# abs function is sqrt (sum of squares of 2 numbers) or distance from origin (0,0) to point a,b
print( np.sqrt( ((1+3j).real) **2 + ((1+3j).imag)**2))

1 <class 'int'>
1j <class 'complex'>
(1+3j) <class 'complex'>
1.0 <class 'float'>
3.0 <class 'float'>
3.1622776601683795 <class 'float'> 

3.1622776601683795


# Topic #5: "C++ vs Python and astronomy applications"

 My understanding is that most astronomers work in Python. 
 
- Python: fastest to write, huge scientific ecosystem (NumPy, Astropy, SciPy, Matplotlib). Ideal for analysis, plotting, file I/O (FITS, CSV), quick prototypes, pipelines, and most production data work.

- C++: fastest at raw computation, full control over memory, great for tight inner loops and heavy simulations. Ideal when you’ve profiled your Python and found true hotspots that need compiled speed.

### Middle ground? Cython
- Cython is the “middle road” between Python and C/C++. You write mostly-Python syntax, add a few static types, and Cython compiles it into a CPython extension module (C under the hood). Result: Python ergonomics, C-like speed on hot loops. It’s perfect to illustrate the trade-offs:

Why it’s great: keep Python code structure; accelerate only hotspots; tight NumPy integration (zero-copy buffers); simpler than writing/maintaining C++ and pybind11 glue.

Why it’s not magic: you still need a C/C++ compiler; builds can be finicky (wheels, ABIs); you must type the bottleneck code and avoid Python objects in inner loops; it’s CPython-specific; ultimate peak performance or advanced C++ features (templates, SIMD libs) still favor pure C++.


### Let's test Python vs Cython (because Dax doesn't know C++... yet!)

#### Cython INSTALLATION
If using Mini-Forge:
```bash
conda install cython
```
If using Anaconda:
```bash
conda install -c conda-forge cython
```

### In this example, we'll do some boxcar smoothing and compare runtimes



In [9]:
# Python code:
def boxsmooth_py(x, w):
    """
    Compute a 1-D boxcar (moving average) over `x` with window width `w`
    using an O(n) rolling-sum update (no re-summing each window).

    Parameters
    ----------
    x : array_like
        1-D sequence of numeric values. Converted to float64 with
        ``np.asarray(x, dtype=float)``.
    w : int
        Boxcar width. Must satisfy ``1 <= w <= len(x)``.

    Returns
    -------
    out : numpy.ndarray
        1-D array of shape ``(len(x) - w + 1,)`` with the moving averages.
        Each output value is aligned to the **right** edge of its window:
        ``out[k] == mean(x[k : k + w])``.

    Raises
    ------
    ValueError
        If ``w <= 0`` or ``w > len(x)``.

    Notes
    -----
    - NaNs inside any window will propagate into the corresponding output
      element; pre-clean or mask as needed.
    - Numerically equivalent (within floating-point tolerance) to
      ``np.convolve(x, np.ones(w)/w, mode='valid')``.
    - Accumulates in float64.

    Examples
    --------
    >>> boxsmooth_py([1, 2, 3, 4], 2)
    array([1.5, 2.5, 3.5])
    """
    x = np.asarray(x, dtype=float)
    if w <= 0 or w > x.size:
        raise ValueError("bad window")
    out = np.empty(x.size - w + 1, dtype=float)
    s = float(np.sum(x[:w])); out[0] = s / w
    for i in range(w, x.size):
        s += x[i] - x[i - w] # Python-level indexing + Python arithmetic
        out[i - w + 1] = s / w # Python-level division + NumPy scalar assign
    return out


### Each iteration does multiple Python operations: dynamic indexing into a NumPy array (returns a Python/NumPy scalar object), dynamic type checks, Python additions/subtractions/division, and a Python→NumPy assignment. Do that ~5 million times and you’re dominated by interpreter overhead, not math.

In [10]:
%load_ext Cython

In [11]:
%%cython
# cython: language_level=3, boundscheck=False, wraparound=False, initializedcheck=False, nonecheck=False
import numpy as np
cimport cython

@cython.boundscheck(False)
@cython.wraparound(False)
def boxsmooth_cy(double[:] x, int w):
    """
    Cython-accelerated 1-D boxcar (moving average) over `x` with window width `w`,
    implemented with an O(n) rolling-sum update.

    Parameters
    ----------
    x : double[:]
        1-D typed memoryview over data convertible to float64 (e.g., a NumPy array).
        Non-float64 or non-contiguous inputs may be copied by the generated wrapper.
        If you require C-contiguous input, use the signature ``double[::1] x``.
    w : int
        Boxcar width. Must satisfy ``1 <= w <= len(x)``.

    Returns
    -------
    numpy.ndarray
        1-D float64 array of shape ``(len(x) - w + 1,)`` containing the moving averages,
        aligned to the **right** edge of each window:
        ``out[k] == mean(x[k : k + w])``.

    Raises
    ------
    ValueError
        If ``w <= 0`` or ``w > len(x)``.

    Notes
    -----
    - Time complexity O(n); extra memory is O(1) beyond the output.
    - Typically much faster than a Python loop; comparable to a well-optimized
      vectorized approach for this operation.
    - NaNs in any window propagate to the corresponding output element.
    - Equivalent to ``np.convolve(x, np.ones(w)/w, mode='valid')`` within FP tolerance.

    Examples
    --------
    >>> import numpy as np
    >>> x = np.arange(5.0)
    >>> boxsmooth_cy(x, 3)
    array([1., 2., 3.])
    """
    cdef Py_ssize_t n = x.shape[0]
    if w <= 0 or w > n:
        raise ValueError("bad window")

    # create a NumPy array (Python object), then view it with a typed memoryview
    y_arr = np.empty(n - w + 1, dtype=np.float64)
    cdef double[:] y = y_arr

    cdef Py_ssize_t i
    cdef double s = 0.0

    # initial sum
    for i in range(w): #initial sum
        s += x[i]
    y[0] = s / w

    # rolling update
    for i in range(w, n):
        s += x[i] - x[i - w] # C loads, C adds/subtracts
        y[i - w + 1] = s / w  # direct write to memory

    return y_arr


#### Why this is faster:

1)  C-level math, not Python objects

    - s is a double, x/y are double[:] memoryviews ⇒ the compiler emits plain floating-point ops. No Python boxing/unboxing.

2) No per-iteration Python dispatch

    - Indexing x[i]/x[i-w] and writing y[...] are pointer arithmetic + bounds (and you disabled bounds/wrap checks), not Python method calls.

3) Removed safety checks

    - boundscheck=False, wraparound=False strips array bounds/negative-index checks from the inner loop.

4) Compiler optimizations

    - -O3 lets the C compiler schedule instructions and sometimes auto-vectorize; Python cannot.

5) Memoryviews keep data contiguous

    - double[:] gives zero-copy, strided access straight to the NumPy buffer; the Python version pays for creating Python scalar objects on every index.

### Benchmark test for runtime comparison

In [12]:
import numpy as np, time, gc

# data outside timing
N = int(5e6)
y = np.sin(np.random.random(N)).astype(np.float64, copy=False)  # contiguous float64
w = 21

# warm up (jit/loader/caches)
# First runs often pay one-time costs: loading code, filling CPU caches, etc.
# Warming up makes the later timings more stable and fair.
_ = boxsmooth_py(y, w)
_ = boxsmooth_cy(y, w)

def bench(fn, x, w, repeats=10):
    """
    Runs any function fn(x, w) several times and returns:
    - the median time (middle value; less sensitive to one weird slow run)
    - the output from the last run (so we can compare results later)

    gc.collect(): ask Python to clean up unused objects before timing, so the garbage collector is less likely to pause during your measurement.
    time.perf_counter(): the right clock for benchmarks (high resolution).
    repeats=10: run 10 times; median gives a robust single number.
    """
    times = []
    for _ in range(repeats):
        gc.collect()
        t0 = time.perf_counter()
        out = fn(x, w)
        t1 = time.perf_counter()
        times.append(t1 - t0)
    return np.median(times), out

t_py, y1 = bench(boxsmooth_py, y, w)
t_cy, y2 = bench(boxsmooth_cy, y, w)

print(f"Pure Python (median of 10): {t_py:.3f} s")
print(f"Cython      (median of 10): {t_cy:.3f} s")
print("Speedup:", round(t_py / t_cy, 1), "×")
print("Max abs diff:", np.max(np.abs(y2 - y1)))


Pure Python (median of 10): 1.231 s
Cython      (median of 10): 0.007 s
Speedup: 173.2 ×
Max abs diff: 2.220446049250313e-16


# Topic #5: "Brief Intro to Bayesian statistics (vs Frequentist)" 

Statistics constitutes an invaluable set of methods and tools for analyzing and making decisions based on data. Their application in various fields has been present for decades or even centuries since the birth of statistics as a discipline in the 18th century.

Traditionally, statistics have been studied and applied by adopting a __frequentist__ approach, based on the idea that the probability of an event is described by the frequency at which that event happens after a large number of experiments or trials.

However, there is a equally powerful approach to statistics: the __Bayesian__ approach. 

Bayesian statistics is just disciplined belief updating.

- You start with a __prior__: your best guess about an unknown quantity before seeing today’s data (e.g., “a coin is probably close to fair”).

- You get data and write a __likelihood__: how compatible the data are with each possible value of the unknown.

- You update your prior with the data to get a __posterior__: your new, sharper belief after seeing the evidence.

- You can then __predict__ new outcomes using the posterior (posterior predictive).

Two things make this useful:

1) It gives you uncertainty as a full distribution, not just a single number.

2) Its intervals (credible intervals) are literal: “there’s a 95% probability the true value lies in this range,” given your model and prior.

$\newline$

Let $\theta$ be the unknown and $D$ be the observed data.

- Prior: p($\theta$)

- Likelihood: p($D | \theta$)

- Posterior (Bayes’ rule):

#### $\large$ p($\theta | D$) =  $\frac{p(D | \theta) p(\theta)}{p(D)}$

where
#### p(D) = $\int~p(D | \theta)~p(\theta)~d\theta$


- Posterior predictive for a future observation y:
#### p(y | D) = $\int~p (y | \theta)~p(\theta | D)~d\theta$

- Credible interval (e.g., 95%): find numbers a < b with:
P( a < $\theta$ < b | D ) = 0.95


# Frequentist statistics treats the world’s truth as fixed 
It's a method of statistical inference that defines probability as the long-run frequency of events in repeated experiments. It focuses on observable data and uses it to make inferences about populations or processes, primarily through hypothesis testing and confidence intervals. 


# A tiny concrete example (coin flips)

## Frequentist approach:
- The coin has a fixed but unknown chance of heads, $\theta$.

- You flip the coin $\textit{n}$ times and count heads $\textit{k}$.

- Your best single estimate is the sample __proportion__ $\hat{p}$ = $\frac{k}{n}$

- A 95% __confidence interval (CI)__ is a procedure that, in repeated experiments, produces an interval that covers the true $\theta$ about 95% of the time. It is not “95% probability this one interval contains $\theta$.

- A p-value for “fair coin” ($\theta$ = 0.5) is: if the coin were truly fair, how likely is data as extreme as yours (or more) just by chance?

### Mathematical Form:


- Let $\textit{X}$ be the number of heads in $\textit{n}$ flips:
#### $X~\sim \mathrm{Binomial}(n,\theta)$ and $\hat{p} = \frac{X}{n}$


A reliable 95% CI for a proportion is the Wilson interval, a method for estimating the confidence interval of a population proportion, particularly when dealing with smaller sample sizes or proportions near the boundaries (0 or 1): 

#### denom = $1+\frac{1+z^2}{n}$
#### center = $\frac{\hat{p} + \frac{z^2}{2n}}{denom}$
#### radius = $\frac{z}{denom} \sqrt{ \frac{\hat{p}(1-\hat{p})}{n} + \frac{z^2}{4n^2} }$

where z is a point on a normal distribution that is used to determine if a test statistic is statistically significant. With z~1.96 for 95%. The CI is [center - radius, center + radius] .


## Bayesian approach:
Unknown: $\theta$ = probability of heads.
Data:  k heads in n flips (Binomial likelihood).

Pick a prior probability distribution. The probability of “success” lives in range [0,1]. The [__Beta distribution__](https://en.wikipedia.org/wiki/Beta_distribution) also lives in [0,1]. 

$\theta~\sim~\beta(\alpha, \beta)$


Then the posterior is closed form:
$\theta~|~ D \sim \beta(\alpha + k,~\beta + n - k)$ 
This is your updated belief about how often the coin lands heads after seeing the flips.

# Say you flip n=10 times and get k=7 heads. 

# Frequentist Approach

In [13]:
def frequentist_coin_approx_bayes(k, n, confidence=0.95, null_p=0.5):
    """
    Frequentist summary for a coin flip experiment that mirrors a simple Bayesian (Beta(1,1)) setup.

    Inputs
    ------
    k : int
        Number of heads (successes).
    n : int
        Number of flips (trials). Must be > 0 and k in [0, n].
    confidence : float
        Confidence level for the Wilson interval. Supports 0.90, 0.95, 0.99.
    null_p : float
        Null value for the hypothesis test (default 0.5 for “fair coin”).

    Returns
    -------
    dict with:
      - p_hat:        k/n (plain frequentist point estimate)
      - p_add_two:    (k+1)/(n+2)  # matches Beta(1,1) posterior mean (a.k.a. Laplace / Agresti–Coull)
      - wilson_ci:    (low, high) Wilson confidence interval at 'confidence'
      - p_value_one_sided:  P(X >= k | n, null_p)     # test H0: p <= null_p vs H1: p > null_p
      - p_value_two_sided:  exact two-sided p-value for H0: p = null_p
      - inputs:       echo of (k, n, confidence, null_p)
    """
    import math
    if n <= 0 or k < 0 or k > n:
        raise ValueError("Require 0 <= k <= n and n > 0.")

    # 1) Point estimates
    p_hat = k / n
    # Agresti–Coull estimate:
    p_add_two = (k + 1) / (n + 2)  # same as Beta(1,1) posterior mean
    # a smoothed estimate of the coin’s heads probability. 
    # It pretends you had one extra head and one extra tail before you started.
    # That gentle nudge keeps estimates sane with tiny samples.

    # 2) Wilson CI (good coverage; stays within [0,1])
    z_map = {0.90: 1.6448536269, 0.95: 1.9599639845, 0.99: 2.5758293035}
    if confidence not in z_map:
        raise ValueError("Use confidence in {0.90, 0.95, 0.99} for this simple helper.")
    z = z_map[confidence]

    denom = 1.0 + (z*z)/n
    center = (p_hat + (z*z)/(2*n)) / denom
    radius = (z / denom) * math.sqrt((p_hat*(1 - p_hat)/n) + (z*z)/(4*n*n))
    wilson_low = max(0.0, center - radius)
    wilson_high = min(1.0, center + radius)

    # 3) Exact binomial p-values under H0: p = null_p
    def binom_pmf(i, n, p):
        return math.comb(n, i) * (p**i) * ((1 - p)**(n - i))

    # One-sided (right tail): evidence of p > null_p
    p_value_one_sided = sum(binom_pmf(i, n, null_p) for i in range(k, n + 1))

    # Two-sided: sum probabilities of outcomes as or less likely than observed
    p_obs = binom_pmf(k, n, null_p)
    eps = 1e-15
    p_value_two_sided = sum(binom_pmf(i, n, null_p) for i in range(n + 1)
                            if binom_pmf(i, n, null_p) <= p_obs + eps)

    return {
        "p_hat": p_hat,
        "p_add_two": p_add_two,
        "wilson_ci": (wilson_low, wilson_high),
        "p_value_one_sided": p_value_one_sided,
        "p_value_two_sided": p_value_two_sided,
        "inputs": {"k": k, "n": n, "confidence": confidence, "null_p": null_p}
    }

# ---- Example case (k=7, n=10) ----

out = frequentist_coin_approx_bayes(k=7, n=10, confidence=0.95, null_p=0.5)
print("p_hat (k/n):", round(out["p_hat"], 4))                  
print("p_add_two:", round(out["p_add_two"], 4))
lo, hi = out["wilson_ci"]
print("Wilson 95% CI:", round(lo, 4), "to", round(hi, 4))
print("One-sided p-value (H0: p <= 0.5):", round(out["p_value_one_sided"], 4)) 
print("Two-sided p-value (H0: p = 0.5):", round(out["p_value_two_sided"], 4))  


p_hat (k/n): 0.7
p_add_two: 0.6667
Wilson 95% CI: 0.3968 to 0.8922
One-sided p-value (H0: p <= 0.5): 0.1719
Two-sided p-value (H0: p = 0.5): 0.3438


# Bayesian approach

## With a uniform prior $\beta$(1,1):

In [14]:
def beta_posterior_summary_from_data(k, n, alpha=1.0, beta=1.0, cred=0.95, nsamples=100, seed=0):
    """
    Binomial-Beta update + posterior summaries.

    Inputs
    ------
    k : int        # number of successes (e.g., heads)
    n : int        # number of trials (e.g., flips)
    alpha, beta : float   # prior Beta(alpha, beta); use alpha=beta=1 for uniform
    cred : float   # credible mass for interval (e.g., 0.95)
    nsamples : int # used only if SciPy is not available (Monte Carlo fallback)
    seed : int     # RNG seed for reproducibility (fallback mode)

    Returns
    -------
    dict with:
      - alpha_post, beta_post
      - mean                 (posterior mean of θ)
      - ci                   ((low, high) credible interval)
      - prob_theta_gt_0_5    (posterior P(θ > 0.5))
    """
    a_post = alpha + k
    b_post = beta + (n - k)
    qlo = (1.0 - cred) / 2.0
    qhi = 1.0 - qlo

    # Try exact quantiles via SciPy; otherwise MC with NumPy
    try:
        from math import isfinite
        from scipy.stats import beta as beta_dist
        mean = a_post / (a_post + b_post)
        lo = float(beta_dist.ppf(qlo, a_post, b_post))
        hi = float(beta_dist.ppf(qhi, a_post, b_post))
        prob_gt_half = float(1.0 - beta_dist.cdf(0.5, a_post, b_post))
    except Exception:
        import numpy as np
        rng = np.random.default_rng(seed)
        samples = rng.beta(a_post, b_post, size=nsamples)
        mean = float(samples.mean())
        lo, hi = np.quantile(samples, [qlo, qhi])
        prob_gt_half = float((samples > 0.5).mean())

    return {
        "alpha_post": float(a_post),
        "beta_post": float(b_post),
        "mean": float(mean),
        "ci": (float(lo), float(hi)),
        "prob_theta_gt_0_5": float(prob_gt_half),
    }
r = beta_posterior_summary_from_data(k=7, n=10, alpha=1, beta=1, cred=0.95)
print('alpha_post:', r['alpha_post'])
print('beta_post:', r['beta_post'])
print('mean:', np.round(r['mean'],3))
print('confidence interval:', r['ci'])
print('H0 (theta > 0.5):',np.round(r['prob_theta_gt_0_5'],3))

alpha_post: 8.0
beta_post: 4.0
mean: 0.667
confidence interval: (0.3902574404275788, 0.8907365561809019)
H0 (theta > 0.5): 0.887


Here’s what each field in this output actually means, in plain English, and how to understand it:

- alpha_post, beta_post
Your updated “pseudo-counts” after seeing the data.
Roughly: alpha_post − 1 behaves like prior+observed “heads,” and beta_post − 1 like prior+observed “tails.”
The sum (alpha_post + beta_post) is the effective sample size the posterior is drawing strength from. Bigger sum → tighter conclusions.

- mean
Your single best guess for the true heads probability after updating with the data (and prior).
Also equals your best predictive probability that the next flip is heads.

- ci (e.g., (0.390, 0.891) for a 95% credible interval)
“Given what we assumed and observed, there’s a 95% chance the true heads probability lies between these two numbers.”
Wider when you have little data; narrower as you collect more.
Practical read: if this interval does not include 0.5, you have strong evidence the coin is not fair (under your model/prior).

- prob_theta_gt_0_5
The probability—after seeing the data—that the coin is more likely than not to land heads (i.e., θ > 0.5).

- Handy decision dial:

    - ~0.5 → we’re basically unsure.
    
    - ~0.7–0.9 → suggestive evidence of heads-bias.
    
    - 0.95 → strong evidence of heads-bias (again, under your model/prior).


## Example: interpreting {'alpha_post': 8, 'beta_post': 4, 'mean': 0.667, 'ci': (0.390, 0.891), 'prob_theta_gt_0_5': 0.887}

- Posterior belief is concentrated toward heads (8 vs 4).

- Best guess: the coin lands heads about 67% of the time.

- Plausible range: true heads rate is likely between 39% and 89% (still wide → not many flips).

- Odds it favors heads:

    - Given your data and (uniform) prior, there’s an 89% chance the true heads-probability is above 0.5.
    - A 95% credible interval is the range that contains the true heads-probability with 95% probability (under your model/prior). Yours is roughly [0.39, 0.89]. Because 0.5 is inside that range, you can’t claim with 95% certainty the coin is more likely than not to land heads. You need more data (or a stronger prior) to tighten that interval so its lower end rises above 0.5.

# Comparing outcomes and what specific questions are being asked

## Bayesian posterior probability:
given the uniform prior Beta(1,1) and the data (7/10), there’s an ~88.7% chance the true long-run heads rate θ exceeds 0.5. That’s a statement about $\theta$.



# Frequentist p-values:

- Two-sided p-value = 0.3438 for 
$(H_0: p=0.5)$:

If the coin were exactly fair, getting a result at least as extreme as 7/10 (too many heads or too many tails) happens ~34% of the time. 

- One-sided p-value = 0.1719 for 
$(H_0: p \le 0.5)$ vs $H_1: p>0.5$:
If the coin were fair or tails-leaning, getting $\ge$7 heads in 10 occurs ~17% of the time. Still not small at 5%, so not significant.

Why can Bayes say “~89% $\theta \gt$0.5” while frequentist p-values aren’t small?

They answer different questions.

Bayes (0.887): probability that the parameter exceeds 0.5 given __data + prior__.

Frequentist p: probability of getting data this extreme or more given the __null__.

With n=10, the data just aren’t extreme enough to make “fair” look rare under repeated sampling. Hence large p-values.

Your 95% credible interval still includes 0.5, which is why, even in the Bayesian framing, you can’t make a 95%-level claim that $\theta \gt$0.5. (You can make an ~89%-level claim.)

Quick translation guide for your numbers (k=7, n=10, Beta(1,1)):

95% credible interval ≈ [0.390, 0.891]: includes 0.5 → not a 95% claim.

Two-sided p = 0.3438: data are common under a fair coin; no rejection.

One-sided p = 0.1719: some heads-leaning, but not significant at 5%.
